ENH: add `out` parameter to arange

csachs opened the pull request:
This PR adds an `out` keyword parameter to the `np.arange` function, to provide an array to be filled in-place with the range.
Looking around I did not spot any other way to perform this operation without allocating new arrays and copying them.

Maybe someone more habituated with Python's refcounting could check the C code for proper refcounting behavior. Thanks.

Example: 
```python
import numpy as np
arr = np.zeros(10)
print(arr)
# [0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]
np.arange(0, len(arr), out=arr)
print(arr)
# [0. 1. 2. 3. 4. 5. 6. 7. 8. 9.]
```

csachs committed:
ENH: add `out` parameter to arange

charris commented:
New features should be run past the mailing list for discussion.

csachs commented:
Sorry, I thought this was only supposed to be done for larger features.
Having all functions be able to work in-place if possible seems desirable to me.

mattip commented:
Couldn't this be spelled as:
```
arr = np.zeros(10, dtype=np.int8)
arr[:] = range(10)
```

seberg commented:
I suspect you would need to spell it with `out` to actually save memory (or do some more manual work).  The way I can think of spelling it would be:
```
out = np.empty(100, dtype=np.intp)
np.add.accumulate(np.broadcast_to(np.array(1, dtype=np.intp), out.shape), out=out)
```
Not 100% sure it won't do unnecessary copies, but I suspect not.

I am OK with the addition (the PR needs some iteration though on first sight).  I am curious if you have a clear use-case @csachs?

charris commented:
Needs rebase and release note.

csachs commented:
Sorry for the delay.
@charris I've rebased it and added a release note, while the newer `PyArray_ArangeObj` code is truely more straightfoward, I hope I now properly added the `out=` changes to it. 
@seberg / @mattip concerning alternatives for this or the reasoning for adding an `out=` argument:
Yes, it's speed, and issues which arise from allocating larger arrays e.g. repeatedly fast. Having `arange` results is pretty common thing, e.g. to quickly create a lot of indices for further manipulation or as a starting point for calculations of function values.
For small arrays it might not make much of an overall impact, but with growing array size, allocating a temporary array can unnecessarily slow down routines.
C.f. the following test:

```python
import numpy as np
blocksize = 1_000_000
dtype = np.uint64
```

```python
%%timeit
arr = np.arange(0, blocksize, dtype=dtype)
#    601 µs ± 36.3 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```
```python
existing_arr = np.zeros(blocksize, dtype=dtype)
```

```python
%%timeit
existing_arr[:] = range(blocksize)
#    65.2 ms ± 335 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```
```python
%%timeit
existing_arr[:] = np.arange(0, blocksize, dtype=dtype)
#    828 µs ± 6.71 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```
```python
%%timeit
np.arange(0, blocksize, out=existing_arr)
#    375 µs ± 1.98 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

Using `range()` is -- compared to the other variants -- tremendously slow (over 150x compared to the fastest), using `arange` and copying the result has expected performance characteristics, but the downside of causing an unnecessary allocation.

seberg commented on numpy/core/src/multiarray/ctors.c:
```diff
@@ -3343,8 +3347,22 @@ PyArray_ArangeObj(PyObject *start, PyObject *stop, PyObject *step, PyArray_Descr
     }
 
+    if (out != Py_None) {
+        if (length != PyArray_DIMS((PyArrayObject *)out)[0]) {
```
Doesn't check for higher dimensions or writeable.

seberg reviewed (changes requested):
I had some basic comments made long ago, but not submitted.  I don't mind the addition, but arange is a bit of a mess that needs cleaning up which makes it harder.

It should get a note to the list as asked for, and needs quite a lot of work to make correct (and new tests).

Thanks for opening and sorry that this is a bit tricky and stalled.  Unless we follow up on this in the next weeks, I suggest closing it.   It can be picked up and the start may be useful for that even if closed.

seberg commented:
@csachs do you have time/energy/motivation to work on this in the next months?  Otherwise, I am tempted to close this as a potentially good idea and start (we can open an issue maybe).  Since it needs some discussion maybe and also quite a bit of work to get right.

csachs commented:
I'm currently on vacation without a laptop; I didn't have time to look into your comments yet; I'd try to look into updating the PR in the next weeks, but if it's preferred to have it closed, we could close it and re-open it when an updated branch is ready.

seberg commented:
Closing, as it's been idle a while and code diverged.  Please don't hesitate to a start a new PR.
